crates/sui-bridge/src/server/handler.rs

```diff
@@ -87,18 +255,202 @@ impl BridgeRequestHandlerTrait for BridgeRequestHandler {
+
+        // TODO: test BridgeEventInUnrecognizedSuiPackage, SuiBridgeEvent::try_from_sui_event
+        // and BridgeEventNotActionable to be cached
+
+        // Test `sign` caches Ok result
+        let emitted_event_1 = MoveTokenBridgeEvent {
+            message_type: BridgeActionType::TokenTransfer as u8,
+            seq_num: 1,
+            source_chain: BridgeChainId::SuiLocalTest as u8,
+            sender_address: SuiAddress::random_for_testing_only().to_vec(),
+            target_chain: BridgeChainId::EthLocalTest as u8,
+            target_address: EthAddress::random().as_bytes().to_vec(),
+            token_type: TokenId::USDC as u8,
+            amount: 12345,
+        };
+
+        // TODO: remove once we don't rely on env var to get package id
+        std::env::set_var("BRIDGE_PACKAGE_ID", "0x0b");
+        init_all_struct_tags();
+
+        let mut sui_event_1 = SuiEvent::random_for_testing();
+        sui_event_1.type_ = SuiToEthTokenBridgeV1.get().unwrap().clone();
+        sui_event_1.bcs = bcs::to_bytes(&emitted_event_1).unwrap();
+        let sui_tx_digest = sui_event_1.id.tx_digest;
+        sui_client_mock.add_events_by_tx_digest(sui_tx_digest, vec![sui_event_1.clone()]);
+        sui_signer_with_cache
+            .sign((sui_tx_digest, sui_event_idx))
+            .await
+            .unwrap();
+
+        // Because the result is cached now, the verifier should not be called again.
+        // Even though we remove the `add_events_by_tx_digest` mock, we will still get the same result.
+        sui_client_mock.add_events_by_tx_digest(sui_tx_digest, vec![]);
+        sui_signer_with_cache
+            .sign((sui_tx_digest, sui_event_idx))
+            .await
+            .unwrap();
+    }
+
+    #[tokio::test]
+    async fn test_eth_signer_with_cache() {
+        let (_, kp): (_, BridgeAuthorityKeyPair) = get_key_pair();
+        let signer = Arc::new(kp);
+        let eth_mock_provider = EthMockProvider::default();
+        let contract_address = EthAddress::random();
+        let eth_client = EthClient::new_mocked(
+            eth_mock_provider.clone(),
+            HashSet::from_iter(vec![contract_address]),
+        );
+        let eth_verifier = EthActionVerifier {
+            eth_client: Arc::new(eth_client),
+        };
+        let mut eth_signer_with_cache = SignerWithCache::new(signer.clone(), eth_verifier);
+
+        // Test `get_cache_entry` creates a new entry if not exist
+        let eth_tx_hash = TxHash::random();
+        let eth_event_idx = 42;
+        assert!(eth_signer_with_cache
+            .get_testing_only((eth_tx_hash, eth_event_idx))
+            .await
+            .is_none());
+        let entry = eth_signer_with_cache
+            .get_cache_entry((eth_tx_hash, eth_event_idx))
+            .await;
+        let entry_ = eth_signer_with_cache
+            .get_testing_only((eth_tx_hash, eth_event_idx))
+            .await;
+        assert!(entry_.unwrap().lock().await.is_none());
```

Comment:
Can we add an assert to explicitly say that there is an entry with that key but that entry is None, for readability? Same for the sui test.
```suggestion
        assert!(entry_.is_some());
        assert!(entry_.unwrap().lock().await.is_none());
```

Reply:
I added a comment